### DISCUSSION TITLE
EntityFilterType - filter by Entitie's field using TextType::class

### DISCUSSION THREAD

**Discussion by Pyrol** [Kreyu/data-table-bundle#123] [General]
Hi, 

Is it possible to filter table by Entity's field with text instead of using ChoiceType?

Primary goal is performance. I'd like to simply use EntityFilterType but with `'form_type' => TextType::class` instead of `ChoiceType::class` because I have 10,000+ entities and it takes some time to grab it from the server everytime I load table.

I tried to use:
 `'query_path' => 'contactPerson'` - filter only works when I put ID of the contactPerson
 `'query_path' => 'contactPerson.surname'` - it says Error: 'contactPerson' is not defined.

My filter is set to entity (ContactPerson)
 
```php
class Ticket{

#[ORM\ManyToOne(inversedBy: 'tickets')]
private ?ContactPerson $contactPerson = null;

}
````
```php
class ContactPerson
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: 'integer')]
    private ?int $id = null;

    #[ORM\Column(length: 64)]
    private ?string $surname = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getSurname(): ?string
    {
        return $this->surname;
    }
}
```

```php
class DataTableTicketType extends AbstractDataTableType
{
    public function buildDataTable(DataTableBuilderInterface $builder, array $options): void
    {
        ->addFilter('contactPerson', EntityFilterType::class, [
            'form_type' => TextType::class,
            'query_path' => 'contactPerson'  //It works ONLY when I put contactPerson's ID in the textfield
            'query_path' => 'contactPerson.surname'  // 'contactPerson.surname': Error: 'contactPerson' is not defined.
        ])
        ;
    }
    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => Ticket::class,
        ]);
    }
}
```

**Comment by Kreyu** [comment]
Hey. As you've noticed, the `EntityFilterType`, as well as the `EntityType` (form type) operates on the entity identifiers. It seems that what you need is a `StringFilterType`, because you want to perform filtering on string/text column. 

```php
use Kreyu\Bundle\DataTableBundle\Bridge\Doctrine\Orm\Filter\Type\StringFilterType;

$builder
  ->addFilter('contactPerson', StringFilterType::class, [
    'query_path' => 'contactPerson.surname',
  ])
;
```

This will also fail with "Error: 'contactPerson' is not defined." if you don't add `contactPerson` alias to the query:

```php
$this->createQueryBuilder('ticket')
  ->addSelect('contactPerson')
  ->leftJoin('ticket.contactPerson', 'contactPerson')
  // ...
;
```

If you used a different alias, for example `person`, then `query_path` should equal to `person.surname`.

This is because the filter adds a WHERE clause to the query like so: `->andWhere('contactPerson.surname LIKE %uniqueparam%')`, so without the join, the Doctrine has no idea about the `contactPerson` alias.

Also, please note that the [`StringFilterType` operator defaults to `Operator::Contains`](https://data-table-bundle.swroblewski.pl/reference/types/filter/doctrine-orm/string.html#default-operator), so the Doctrine applies a LIKE instead of an exact match. If you need exact match, [change `default_operator` option to `Operator::Equals`](https://data-table-bundle.swroblewski.pl/docs/components/filters.html#changing-default-operator).

Hope that helps.

**Comment by Pyrol** [comment]
Works as expected <3. 

Example in controller for future symonfy devs :) 

SiteController.php
```php
    #[Route('/path/to/get/ticket/table', name: 'get_ticket_datatable',  methods: ['GET'])]
    public function get_ticket_datatable(Request $request, TicketRepository $ticketRepository): Response
    {

        $qb = $ticketRepository->createQueryBuilder('ticket')    
        ->addSelect('contactPerson')
        ->leftJoin('ticket.contactPerson', 'contactPerson');

        $dt = $this->createDataTable(DataTableTicketType::class,$qb);     
        $dt->handleRequest($request);

        return new Response($this->renderView('/dataTable/_dataTable.html.twig', [
            'dt' => $dt->createView(),
            'target' => 'ticketTable'
        ]),
        200 ,
        ['Content-Type' => 'text/vnd.turbo-stream.html']);

    }
```
